File: backend/models/forum.py
from app import db
from datetime import datetime

class ForumThread(db.Model):
    __tablename__ = 'forum_threads'
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    title = db.Column(db.String(255), nullable=False)
    content = db.Column(db.Text, nullable=False)
    semester = db.Column(db.String(50), nullable=False, index=True)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    user = db.relationship('User', backref='threads')
    replies = db.relationship('ForumReply', backref='thread', cascade='all, delete-orphan')

    def to_dict(self):
        return {
            'id': self.id,
            'userId': self.user_id,
            'userName': self.user.email,
            'title': self.title,
            'content': self.content,
            'semester': self.semester,
            'createdAt': self.created_at.isoformat(),
            'replyCount': len(self.replies)
        }

class ForumReply(db.Model):
    __tablename__ = 'forum_replies'
    id = db.Column(db.Integer, primary_key=True)
    thread_id = db.Column(db.Integer, db.ForeignKey('forum_threads.id'), nullable=False)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    content = db.Column(db.Text, nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    user = db.relationship('User', backref='replies')

    def to_dict(self):
        return {
            'id': self.id,
            'threadId': self.thread_id,
            'userId': self.user_id,
            'userName': self.user.email,
            'content': self.content,
            'createdAt': self.created_at.isoformat()
        }

File: lib/data/forum.ts
import "server-only";
import { db } from "@/lib/prisma";
import { parseId } from "@/lib/parseId";
import { canAccessSemester } from "@/lib/auth/dal";
import type { User, ForumThread, ForumReply } from "@/generated/prisma/client";

function threadToDict(t: ForumThread & { user: User; _count: { replies: number } }) {
  return {
    id: t.id,
    userId: t.userId,
    userName: t.user.email,
    title: t.title,
    content: t.content,
    semester: t.semester,
    createdAt: t.createdAt.toISOString(),
    replyCount: t._count.replies,
  };
}

function replyToDict(r: ForumReply & { user: User }) {
  return {
    id: r.id,
    threadId: r.threadId,
    userId: r.userId,
    userName: r.user.email,
    content: r.content,
    createdAt: r.createdAt.toISOString(),
  };
}

export type ThreadListItem = ReturnType<typeof threadToDict>;
export type ReplyItem = ReturnType<typeof replyToDict>;

// Forum threads are scoped to the viewer's own semester — including for
// admins. The original UI never exposed a cross-semester filter here
// (unlike the PDF dashboard), so there's nothing to replicate beyond this.
export async function listThreadsForUser(user: User, page: number, perPage: number) {
  const safePage = Math.max(page || 1, 1);
  const safePerPage = Math.min(Math.max(perPage || 10, 1), 50);

  const [threads, total] = await Promise.all([
    db.forumThread.findMany({
      where: { semester: user.semester },
      include: { user: true, _count: { select: { replies: true } } },
      orderBy: { createdAt: "desc" },
      skip: (safePage - 1) * safePerPage,
      take: safePerPage,
    }),
    db.forumThread.count({ where: { semester: user.semester } }),
  ]);

  return {
    threads: threads.map(threadToDict),
    total,
    pages: Math.ceil(total / safePerPage),
    currentPage: safePage,
  };
}

export async function getThreadForUser(rawId: string, user: User) {
  const id = parseId(rawId);
  if (id === null) return null;

  const thread = await db.forumThread.findUnique({
    where: { id },
    include: {
      user: true,
      _count: { select: { replies: true } },
      replies: { include: { user: true }, orderBy: { createdAt: "asc" } },
    },
  });
  if (!thread) return null;
  if (!canAccessSemester(user, thread.semester)) return null;

  return { thread: threadToDict(thread), replies: thread.replies.map(replyToDict) };
}
